awslabs/amazon-kinesis-video-streams-webrtc-sdk-js
# Sample: JoinStorageSession retries and usability enhancements

## What was changed?
* Every WebRTC session from now on will now start with a describeMediaStorageConfiguration. We need to use the result of this to determine which mode to run the sample application in.
* Add JoinStorageSession retries. This only applies in When in _Ingestion & Storage_ mode.
  * If no SDP offer was received over the WebSocket within 6 seconds, we try again, with some backoff and jitter.
  * If we receive SDP offers from the _KVS Ingestion & Storage peer_ within a 10-minute window, and the peer connection failed to be established 5 times within that 10-minute window, we will exit the application. The user would need to manually restart the sample in that case.
  * When the connection to the _KVS Ingestion & Storage peer_ is lost, we will attempt to reconnect by making a JoinStorageSession API call. Due to the [1-hour session limit](https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-limits.html), we will lose connection by that point, and attempt to connect again.
  * Due to the [WebSocket closing after 10 minutes of inactivity](https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-limits.html), we need to reconnect to the WebSocket whenever we make the JoinStorageSession API call, if it has been closed.
<img width="925" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/8ebdaa77-2080-462f-8220-487322e0393c">
<img width="1111" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/1ceac7cd-a023-4023-9d33-1a82566d5b53">

* Rename WebRTC Storage section to WebRTC Storage and Ingestion
* Make the WebRTC Storage and Ingestion section hidden by default, can expand the section to show the options.
* Add a button allowing users to configure which stream to ingest and store the media to:
<img width="961" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/99ce550d-6f54-4b92-84a9-0664778bd63d">

* Add a button to print what is the current mediaStorageConfiguration for the signaling channel.
 
<img width="331" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/96a1ad5f-01ce-4276-bf00-995a24db4876">
<img width="1077" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/ef9a1c7d-e339-4506-8db0-9d0e71e60838">

* Print selected options to the logs when the sample is started.

<img width="820" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/651a1ccf-b1c7-4beb-ad90-a6ff29263e7a">

* Rename the "Ingest media" checkbox, to "Ingestion and storage peer joins automatically". The behavior of this button has been tweaked to have no effect (i.e. don't call JoinStorageSession) when the media storage configuration is DISABLED. This checkbox is also enabled by default.
  * As a result, the "Audio and video" checkboxes needing to be checked for WebRTC ingestion and storage has been moved to an application _runtime_ check.

<img width="379" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/b6f8dcea-67a6-4eaf-8ff7-600c3fc80f12">

* Add tooltip for the "Ingestion and storage peer joins automatically" checkbox.
<img width="255" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/aaabb9ac-d89a-4259-9498-1991fa54098f">

* Since regular viewers can no longer join, when this feature is enabled, add an error message when the viewer tries to connect to a channel with media storage is enabled (prevent failing silently).
<img width="1088" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/889d7393-e5c0-4401-b85a-ea4126bacdee">

* (By @stefankiesz) add a button to print all of the channels currently configured for ingestion.

<img width="756" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/ec0b629d-2710-4999-ac83-a04923f68750">
<img width="1059" alt="image" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/14988194/1bd36930-85b8-4651-8d95-d28b77991e95">

## Why was it changed?
* In case JoinStorageSession is called and KVS backend is experiencing problems, retry the call in order to receive an SDP offer from the _Ingestion and Storage Peer_.
* The 5 failure attempts within 10 minutes was added in order to prevent the following _infinite_ loop, where we will continuously attempt to establish a connection:
1. Call JoinStorageSession
2. Receive SDP offer
3. Send Answer, exchange ice candidates, etc
4. Peer connection failed to be established (e.g. due to firewall)
5. Call JoinStorageSession, since the peer connection failed.
6. Receive SDP offer
7. And so on.
* The rest of the changes were made to enhance the user experience. Needing to use the AWS CLI to configure the media storage configuration made that part not "fit" with the rest of the sample (where everything can be done on the web).
* WebRTC Ingestion & Storage section was made hidden by default because it adds a bunch of clutter. Users using the regular WebRTC will have no need for this.

## Testing
* Manually tested
  * Tested reconnect by running sample for multiple hours, and also randomly disconnecting from internet until the log message "connection peer to failed" shows up before connecting to internet again.
  * Tested exponential backoff/jitter by putting a return; as the first thing in SiganlingClient.ts onMessage
  * Tested 5 retries within 10 minutes by disabling STUN and TURN.
  * Update media storage configuration button with empty streamname, and to a new streamname.
  * Click describe storage configuration button, the output is as expected.
  * Click the list channels button, confirm that the output is as expected.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
